File: packages/compiler/src/transforms/vFor.ts
import { DynamicFlag, IRNodeTypes, type SimpleExpressionNode } from '../ir'
import {
  createStructuralDirectiveTransform,
  type NodeTransform,
  type TransformContext,
} from '../transform'
import {
  createBranch,
  createCompilerError,
  EMPTY_EXPRESSION,
  ErrorCodes,
  findProp,
  getExpression,
  isConstantNode,
  isEmptyText,
  isJSXComponent,
  resolveExpression,
} from '../utils'
import type { JSXAttribute, JSXElement } from 'oxc-parser'

export const transformVFor: NodeTransform = createStructuralDirectiveTransform(
  'for',
  processFor,
)

export function processFor(
  node: JSXElement,
  dir: JSXAttribute,
  context: TransformContext,
) {
  const { value, index, key, source } = getForParseResult(dir, context)
  if (!source) {
    context.options.onError(
      createCompilerError(ErrorCodes.X_V_FOR_MALFORMED_EXPRESSION, dir.range),
    )
    return
  }

  const keyProp = findProp(node, 'key')
  const keyProperty =
    keyProp?.type === 'JSXAttribute' && keyProp.value
      ? resolveExpression(keyProp.value, context)
      : undefined
  const isComponent =
    isJSXComponent(node) ||
    // template v-for with a single component child
    isTemplateWithSingleComponent(node)
  const id = context.reference()
  context.dynamic.flags |= DynamicFlag.NON_TEMPLATE | DynamicFlag.INSERT
  const [render, exitBlock] = createBranch(node, context, true)
  return (): void => {
    exitBlock()

    const { parent } = context

    // if v-for is the only child of a parent element, it can go the fast path
    // when the entire list is emptied
    const isOnlyChild =
      parent &&
      parent.block.node !== parent.node &&
      parent.node.children.filter((child) => !isEmptyText(child)).length === 1

    context.dynamic.operation = {
      type: IRNodeTypes.FOR,
      id,
      source,
      value,
      key,
      index,
      keyProp: keyProperty,
      render,
      once: context.inVOnce || !!(source.ast && isConstantNode(source.ast)),
      component: isComponent,
      onlyChild: !!isOnlyChild,
    }
  }
}

export function getForParseResult(
  dir: JSXAttribute,
  context: TransformContext,
) {
  let value: SimpleExpressionNode | undefined,
    index: SimpleExpressionNode | undefined,
    key: SimpleExpressionNode | undefined,
    source: SimpleExpressionNode | undefined
  if (dir.value) {
    const expression =
      dir.value.type === 'JSXExpressionContainer'
        ? getExpression(dir.value)
        : undefined
    if (expression?.type === 'BinaryExpression') {
      const left = getExpression(expression.left)
      if (left.type === 'SequenceExpression') {
        const expressions = left.expressions
        value = expressions[0] && resolveExpression(expressions[0], context)
        key = expressions[1] && resolveExpression(expressions[1], context)
        index = expressions[2] && resolveExpression(expressions[2], context)
      } else {
        value = resolveExpression(left, context)
      }
      source = resolveExpression(expression.right, context)
    }
  } else {
    context.options.onError(
      createCompilerError(ErrorCodes.X_V_FOR_NO_EXPRESSION, dir.range),
    )
  }

  return {
    value,
    index,
    key,
    source,
  }
}

function isTemplateWithSingleComponent(node: JSXElement): boolean {
  const nonCommentChildren = node.children.filter((c) => !isEmptyText(c))
  return (
    nonCommentChildren.length === 1 && isJSXComponent(nonCommentChildren[0])
  )
}

File: packages/compiler-rs/src/transform/v_for.rs
use napi::{
  Either,
  bindgen_prelude::{Either3, Either16},
};
use oxc_allocator::TakeIn;
use oxc_ast::ast::{
  BinaryExpression, Expression, JSXAttribute, JSXAttributeValue, JSXChild, JSXElement,
};

use crate::{
  ir::index::{BlockIRNode, DynamicFlag, ForIRNode, IRFor, SimpleExpressionNode},
  transform::{ContextNode, TransformContext},
  utils::{
    check::{is_constant_node, is_jsx_component, is_template},
    error::ErrorCodes,
    text::is_empty_text,
    utils::{find_prop, find_prop_mut},
  },
};

pub fn transform_v_for<'a>(
  context_node: *mut ContextNode<'a>,
  context: &'a TransformContext<'a>,
  context_block: &'a mut BlockIRNode<'a>,
  parent_node: &'a mut ContextNode<'a>,
) -> Option<Box<dyn FnOnce() + 'a>> {
  let Either::B(JSXChild::Element(node)) = (unsafe { &mut *context_node }) else {
    return None;
  };
  let node = node as *mut oxc_allocator::Box<JSXElement>;
  if is_template(unsafe { &*node })
    && find_prop(unsafe { &*node }, Either::A("v-slot".to_string())).is_some()
  {
    return None;
  }

  let Some(dir) = find_prop_mut(unsafe { &mut *node }, Either::A("v-for".to_string())) else {
    return None;
  };
  let seen = &mut context.seen.borrow_mut();
  let span = dir.span;
  if seen.contains(&span.start) {
    return None;
  }
  seen.insert(span.start);

  let Some(IRFor {
    value,
    index,
    key,
    source,
  }) = get_for_parse_result(dir, context)
  else {
    return None;
  };

  let Some(source) = source else {
    context.options.on_error.as_ref()(ErrorCodes::VForMalformedExpression, span);
    return None;
  };

  let key_prop = if let Some(key_prop) =
    find_prop_mut(unsafe { &mut *node }, Either::A("key".to_string()))
    && let Some(value) = &mut key_prop.value
  {
    Some(SimpleExpressionNode::new(Either3::C(value), context))
  } else {
    None
  };

  let component =
    is_jsx_component(unsafe { &*node }) || is_template_with_single_component(unsafe { &*node });
  let dynamic = &mut context_block.dynamic;
  let id = context.reference(dynamic);
  dynamic.flags = dynamic.flags | DynamicFlag::NonTemplate as i32 | DynamicFlag::Insert as i32;
  let block = context_block as *mut BlockIRNode;
  let exit_block = context.create_block(
    unsafe { &mut *context_node },
    unsafe { &mut *block },
    Expression::JSXElement(oxc_allocator::Box::new_in(
      unsafe { &mut *node }.take_in(context.allocator),
      context.allocator,
    )),
    Some(true),
  );

  // if v-for is the only child of a parent element, it can go the fast path
  // when the entire list is emptied
  let mut only_child = false;
  if let Either::B(JSXChild::Element(parent_node)) = parent_node
    && !is_jsx_component(parent_node)
  {
    let index = *context.index.borrow() as usize;
    for (i, child) in parent_node.children.iter().enumerate() {
      let child = if index == i {
        match unsafe { &mut *context_node } {
          Either::A(_) => child,
          Either::B(node) => node,
        }
      } else {
        child
      };
      if !is_empty_text(child) {
        if only_child {
          only_child = false;
          break;
        }
        only_child = true;
      }
    }
  };

  Some(Box::new(move || {
    let block = exit_block();

    context_block.dynamic.operation = Some(Box::new(Either16::B(ForIRNode {
      id,
      value,
      key,
      index,
      key_prop,
      render: block,
      once: *context.in_v_once.borrow() || is_constant_node(&source.ast.as_deref()),
      source,
      component,
      only_child,
      parent: None,
      anchor: None,
    })));
  }))
}

pub fn get_for_parse_result<'a>(
  dir: &'a mut JSXAttribute<'a>,
  context: &'a TransformContext<'a>,
) -> Option<IRFor<'a>> {
  let mut value: Option<SimpleExpressionNode> = None;
  let mut index: Option<SimpleExpressionNode> = None;
  let mut key: Option<SimpleExpressionNode> = None;
  let mut source: Option<SimpleExpressionNode> = None;
  if let Some(dir_value) = &mut dir.value {
    let expression = if let JSXAttributeValue::ExpressionContainer(dir_value) = dir_value {
      Some(
        dir_value
          .expression
          .to_expression_mut()
          .without_parentheses_mut()
          .get_inner_expression_mut(),
      )
    } else {
      None
    };
    if let Some(expression) = expression
      && let Expression::BinaryExpression(expression) = expression
    {
      let expression = expression as *mut oxc_allocator::Box<BinaryExpression>;
      let left = unsafe { &mut *expression }
        .left
        .without_parentheses_mut()
        .get_inner_expression_mut();
      if let Expression::SequenceExpression(left) = left {
        let expressions = &mut left.expressions as *mut oxc_allocator::Vec<Expression>;
        value = unsafe { &mut *expressions }
          .get_mut(0)
          .map(|e| SimpleExpressionNode::new(Either3::A(e), context));
        key = unsafe { &mut *expressions }
          .get_mut(1)
          .map(|e| SimpleExpressionNode::new(Either3::A(e), context));
        index = unsafe { &mut *expressions }
          .get_mut(2)
          .map(|e| SimpleExpressionNode::new(Either3::A(e), context));
      } else {
        value = Some(SimpleExpressionNode::new(Either3::A(left), context));
      };
      source = Some(SimpleExpressionNode::new(
        Either3::A(&mut unsafe { &mut *expression }.right),
        context,
      ));
    }
  } else {
    context.options.on_error.as_ref()(ErrorCodes::VForNoExpression, dir.span);
    return None;
  }
  return Some(IRFor {
    value,
    index,
    key,
    source,
  });
}

fn is_template_with_single_component<'a>(node: &'a JSXElement<'a>) -> bool {
  let non_comment_children = node
    .children
    .iter()
    .filter(|c| !is_empty_text(c))
    .collect::<Vec<_>>();

  non_comment_children.len() == 1
    && matches!(non_comment_children[0],JSXChild::Element(child)if is_jsx_component(child))
}
